pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1105 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	informers "k8s.io/client-go/informers/core/v1"
+	networkinginformers "k8s.io/client-go/informers/networking/v1alpha1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	networkinglisters "k8s.io/client-go/listers/networking/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/component-base/metrics/prometheus/ratelimiter"
+	nodeutil "k8s.io/component-helpers/node/util"
+	"k8s.io/klog/v2"
+	v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	cidrset "k8s.io/kubernetes/pkg/controller/nodeipam/ipam/multicidrset"
+	cccutil "k8s.io/kubernetes/pkg/controller/util/clustercidrconfig"
+	controllerutil "k8s.io/kubernetes/pkg/controller/util/node"
+	"k8s.io/kubernetes/pkg/util/slice"
+	netutil "k8s.io/utils/net"
+)
+
+const (
+	defaultClusterCIDRConfigKey        = "kubernetes.io/clusterCIDRConfig"
+	defaultClusterCIDRConfigValue      = "default"
+	defaultClusterCIDRConfigName       = "default-cluster-cidr-config"
+	defaultClusterCIDRConfigAPIVersion = "networking.k8s.io/v1alpha1"
+	clusterCIDRConfigFinalizer         = "networking.k8s.io/cluster-cidr-config-finalizer"
+	ipv4MaxCIDRMask                    = 32
+	ipv6MaxCIDRMask                    = 128
+	minPerNodeHostBits                 = 4
+	maxRetries                         = 15
+)
+
+// CIDRs are reserved, then node resource is patched with them.
+// multiCIDRNodeReservedCIDRs holds the reservation info for a node.
+type multiCIDRNodeReservedCIDRs struct {
+	nodeReservedCIDRs
+	clusterCIDR *cidrset.ClusterCIDR
+}
+
+type multiCIDRRangeAllocator struct {
+	client clientset.Interface
+	// nodeLister is able to list/get nodes and is populated by the shared informer passed to controller.
+	nodeLister corelisters.NodeLister
+	// nodesSynced returns true if the node shared informer has been synced at least once.
+	nodesSynced cache.InformerSynced
+	// clusterCIDRConfigLister is able to list/get clustercidrconfigs and is populated by the shared informer passed to controller.
+	clusterCIDRConfigLister networkinglisters.ClusterCIDRConfigLister
+	// clusterCIDRConfigSynced returns true if the clustercidrconfig shared informer has been synced at least once.
+	clusterCIDRConfigSynced cache.InformerSynced
+	// Channel that is used to pass updating Nodes and their reserved CIDRs to the background.
+	// This increases a throughput of CIDR assignment by not blocking on long operations.
+	nodeCIDRUpdateChannel chan multiCIDRNodeReservedCIDRs
+	recorder              record.EventRecorder
+	// Keep a set of nodes that are currently being processed to avoid races in CIDR allocation.
+	lock              *sync.Mutex
```

Comment:
for lock protected fields, these should be offset in their own block so it's clear what fields require holding the lock:

```
struct {
  field1 int
  field2 int
  <<< empty line >>>
  lock ...
  guardedField ..
  guardedField ...
  <<< empty line >>>
  notGuardedField...
  notGuardedField..
}


```


```
lock *sync.Mutex
nodesInProcess sets.String

?? is queue also guarded by the lock

Reply:
Your lock comment should clearly state which fields require the lock to be held.

Reply:
Done